### PR TITLE
Show the number of people having a cache on watchlist (Fix Issue #4058)

### PR DESCRIPTION
**Pull request by ThibaultR**
In CacheDetailActivity you can now see "Cache is not on your watchlist (6 in total)" or "Cache is on your watchlist (7 in total)".

### CODE REVIEW

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 1]
```diff
@@ -1276,14 +1276,21 @@ private void updateWatchlistBox() {
             final Button buttonRemove = ButterKnife.findById(view, R.id.remove_from_watchlist);
             final TextView text = ButterKnife.findById(view, R.id.watchlist_text);
 
+            final int watchListCount = cache.getWatchlistCount();
+            String watchListCountString = "";
+            if (watchListCount != -1)
+            {
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/CacheDetailActivity.java]: Opening brace goes on the same lilne as the `if`.

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 2]
```diff
@@ -1276,14 +1276,21 @@ private void updateWatchlistBox() {
...
+                watchListCountString = " (" + watchListCount + ' ' + getString(R.string.cache_watchlist_in_total) + ')';
+            }
+
             if (cache.isOnWatchlist() || cache.isOwner()) {
                 buttonAdd.setVisibility(View.GONE);
                 buttonRemove.setVisibility(View.VISIBLE);
-                text.setText(R.string.cache_watchlist_on);
+                text.setText(res.getString(R.string.cache_watchlist_on) + watchListCountString);
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/CacheDetailActivity.java]: Please use a single string resource to do this. Different languages may present things in different orders.

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 3]
```diff
@@ -1276,14 +1276,21 @@ private void updateWatchlistBox() {
...
             } else {
                 buttonAdd.setVisibility(View.VISIBLE);
                 buttonRemove.setVisibility(View.GONE);
-                text.setText(R.string.cache_watchlist_not_on);
+                text.setText(res.getString(R.string.cache_watchlist_not_on) + watchListCountString);
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/CacheDetailActivity.java]: Do not repeat yourself, compute it once and for all.

File: main/src/cgeo/geocaching/DataStore.java [hunk 4]
```diff
@@ -1782,6 +1785,8 @@ private static Geocache createCacheFromDatabaseContent(final Cursor cursor) {
         cache.setCoords(getCoords(cursor, 35, 36));
         cache.setFinalDefined(cursor.getInt(37) > 0);
         cache.setLogPasswordRequired(cursor.getInt(41) > 0);
+        cache.setWatchlistCount(cursor.getInt(42));
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/DataStore.java]: You need to increment the database version and add a column in the migration.

File: main/src/cgeo/geocaching/Geocache.java [hunk 5]
```diff
@@ -1092,6 +1093,20 @@ public void setOnWatchlist(final boolean onWatchlist) {
     }
 
     /**
+     *
+     * Set the number of users watching this geocache
+     * @param watchlistCount Number of users watching this geocache
+     */
+    public void setWatchlistCount(final int watchlistCount) {this.watchlistCount = watchlistCount;}
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/Geocache.java]: Function body goes on a different line.

File: main/src/cgeo/geocaching/connector/gc/GCParser.java [hunk 6]
```diff
@@ -461,6 +461,10 @@ static SearchResult parseAndSaveCacheFromText(final String page, @Nullable final
         // cache guid
         cache.setGuid(TextUtils.getMatch(page, GCConstants.PATTERN_GUID, true, cache.getGuid()));
 
+        // cache watchlistcount
+        int watchListCount = getWatchListCount(page);
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/connector/gc/GCParser.java]: Variable should be `final`, or could be inlined.

File: main/src/cgeo/geocaching/connector/gc/GCParser.java [hunk 7]
```diff
@@ -1328,10 +1332,33 @@ static boolean addToWatchlist(final Geocache cache) {
...
+     * @param page Page containing the information about howm many people watching on geocache
+     * @return Number of people watching geocache, -1 when error
+     */
+    static int getWatchListCount(final String page)
+    {
+        int count;
+        String sCount = TextUtils.getMatch(page, GCConstants.PATTERN_WATCHLIST_COUNT, true, 1, "notFound", false);
+        if (sCount.equals("notFound")) return -1;
+        try {
+            count = Integer.parseInt(sCount);
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/connector/gc/GCParser.java]: Just return the result from `parseInt` from here, no need for the `count` variable.

File: main/src/cgeo/geocaching/connector/gc/GCParser.java [hunk 8]
```diff
@@ -1328,10 +1332,33 @@ static boolean addToWatchlist(final Geocache cache) {
...
+        } catch(NumberFormatException nfe) {
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/connector/gc/GCParser.java]: `nfe` should be `final`

File: main/src/cgeo/geocaching/connector/gc/GCParser.java [hunk 9]
```diff
@@ -1328,10 +1332,33 @@ static boolean addToWatchlist(final Geocache cache) {
...
+            Log.i("Could not parse " + nfe);
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/connector/gc/GCParser.java]: This should be an error, and the exception should be the second parameter.

**Comment by samueltardieu** [comment]
Please use the imperative in your commit messages ("Add xxx" instead of "added xxx").

**Comment by Lineflyer** [comment]
@ThibaultR Shall we keep this PR open. Are you working on the review comments?

**Comment by ThibaultR** [comment]
Hi ! 
Sorry for the wait, forgot about it with exam and everything...
I think I have made all the changes required, not sure about the string part and the translation though. Same for the changes in the DataStore...

File: main/src/cgeo/geocaching/CacheDetailActivity.java [hunk 10]
```diff
@@ -1276,14 +1276,20 @@ private void updateWatchlistBox() {
             final Button buttonRemove = ButterKnife.findById(view, R.id.remove_from_watchlist);
             final TextView text = ButterKnife.findById(view, R.id.watchlist_text);
 
+            final int watchListCount = cache.getWatchlistCount();
+            String watchListCountString = "";
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/CacheDetailActivity.java]: Non-`final` variables are error-prone and should be avoided. Could you use something like this instead?

``` java
   final String watchListCountString = watchListCount != -1 ? '(' + … + ')' : "";
```

**Comment by samueltardieu** [comment]
I have merged it with the following changes:
- database version is now 70 (there has been another change in the meantime);
- I have prevented the display of an extra space when we have no watchers information;
- I have invalidated the previous transactions to prevent non-English languages from swallowing the indication.

**Pull request merged**